arch/riscv: support CONFIG_CODE_DATA_RELOCATION

tari opened the pull request:
This implements support for relocating code to chosen memory regions via
the `zephyr_code_relocate` CMake function for RISC-V, where it was
previously implemented only for ARM.

tari committed:
arch/riscv: support CONFIG_CODE_DATA_RELOCATION

This implements support for relocating code to chosen memory regions via
the `zephyr_code_relocate` CMake function for RISC-V SoCs. ARM-specific
assumptions that were made by gen_relocate_app.py need to be corrected,
in particular not assuming any particular name for the default RAM
section (which is 'SRAM' for most ARM pltaforms) and not assuming 32-bit
pointers (so the test works on RV64).

Signed-off-by: Peter Marheine <pmarheine@chromium.org>

tari committed:
build: don't assume .symtab section type

Only the elftools SymbolTableSection section type provides an
iter_symbols() method, and compilers are free to emit sections that
include the substring '.symtab' in their name which causes errors if
gen_relocate_app examines only the section name. Instead check whether a
section is a symbol table to skip attempting to inspect sections that
are not actually symbol tables.

Signed-off-by: Peter Marheine <pmarheine@chromium.org>

tari committed:
build: don't assume FLASH is the ROM region

Most ARM platforms name their ROM region `FLASH`, but this assumption is
not portable. Have gen_relocate_app refer to ROMABLE_REGION instead of
hard-coding the platform-specific memory region name.

Signed-off-by: Peter Marheine <pmarheine@chromium.org>

carlocaione commented on arch/riscv/Kconfig:
```diff
@@ -128,6 +128,24 @@ config GEN_IRQ_VECTOR_TABLE
 
+config CODE_DATA_RELOCATION
+	bool "Support code/data section relocation"
+	depends on SOC_SERIES_RISCV32_IT8XXX2 \
+	  || SOC_RISCV32_LITEX_VEXRISCV \
+	  || SOC_SERIES_GD32VF103 \
+	  || SOC_SERIES_RISCV32_MIV \
+	  || SOC_SERIES_RISCV64_MIV \
+	  || SOC_SERIES_NEORV32 \
+	  || SOC_SERIES_RISCV_SIFIVE_FREEDOM \
+	  || SOC_SERIES_STARFIVE_JH71XX \
+	  || SOC_SERIES_RISCV_TELINK_B91 \
+	  || SOC_SERIES_RISCV_VIRT
```
Uhm, this is not particularly scalable and we will 100% forget to keep this updated with new SoCs coming. Can you instead add something like `SOC_SUPPORTS_RELOC` (please, find a better name :) that each SoC can set at SoC defconfig level and make this symbol depending on that instead?

tari replied:
Added `ARCH_SUPPORTS_CODE_DATA_RELOCATION` at top-level and made all currently-supporting systems select it.

carlocaione commented on arch/riscv/Kconfig:
```diff
@@ -128,6 +128,24 @@ config GEN_IRQ_VECTOR_TABLE
+	  || SOC_SERIES_RISCV_TELINK_B91 \
+	  || SOC_SERIES_RISCV_VIRT
+	help
+	  Enable support for relocating .text, data and .bss sections from specified
+	  files and placing them in a chosen memory region. Files to relocate and
+	  the target regions should be specified in CMakeLists.txt using
+	  zephyr_code_relocate().
```
This is an exact duplicate of the one we have in `arch/arm/Kconfig` already, so why don't simply unify the two symbols in a general one for all the architectures supporting this?

tari replied:
I didn't want to move the ARM version out of `arch/arm` since it has direct dependence on ARM specifics, but if we're adding a new symbol for "relocation is supported" as suggested then it makes sense to move it upward.

carlocaione reviewed (changes requested):
We need also to make this test-able by `tests/application_development/code_relocation/`

fkokosinski commented:
Hi @tari, thanks for your contribution. Did you test these changes on any of the RISC-V boards?

tari commented:
> Hi @tari, thanks for your contribution. Did you test these changes on any of the RISC-V boards?

Yes, I'm testing on it8xxx2.

tari committed:
arch: move CODE_DATA_RELOCATION to top level

Support for CODE_DATA_RELOCATION is not inherently limited to ARM, so
move the Kconfig definition to top-level so it can be used by other
architectures. Since support is opt-in (requiring linker script
support), add a helper symbol enabled by architecture config that gates
whether CODE_DATA_RELOCATION is available instead of listing all
supported systems inline.

Signed-off-by: Peter Marheine <pmarheine@chromium.org>

carlocaione commented on soc/riscv/litex-vexriscv/Kconfig.soc:
```diff
@@ -9,6 +9,7 @@ config SOC_RISCV32_LITEX_VEXRISCV
 	select RISCV_ISA_RV32I
 	select RISCV_ISA_EXT_M
 	select RISCV_ISA_EXT_A
+	select ARCH_SUPPORTS_CODE_DATA_RELOCATION
```
ok, I found this a bit confusing because you are setting an `ARCH_*` property from a SoC configuration file. The `ARCH_HAS / ARCH_SUPPORTS` should only be enabled by the architecture (like the name is suggesting).

`ARCH_HAS_CODE_DATA_RELOCATION` should be enable at arch level in `arch/riscv/Kconfig` (because you indeed modified `include/zephyr/arch/riscv/common/linker.ld`) while at the SoC level you must only set `CONFIG_CODE_DATA_RELOCATION` for the SoCs that actually supports it.

tari replied:
If `ARCH_HAS_CODE_DATA_RELOCATION` is selected at top-level `riscv`, then we still need to express whether a given SoC actually uses `arch/riscv/common/linker.ld`. It seems like we'd need to add a `riscv`-specific `SOC_HAS_CODE_DATA_RELOCATION` symbol that is selected by those SoCs that support it, since `CODE_DATA_RELOCATION` should be prevented from being enabled unless the system actually supports it. I don't mind adding that layer of indirection, but it seems a bit unnecessary only to avoid setting an `ARCH_` symbol at SoC-level (though I grant that's a pretty minor stylistic opinion).

Unless you had something else in mind?

carlocaione replied:
I missed that some SoCs are not actually using `arch/riscv/common/linker.ld` (why thought?), this sucks.

Can you point me to the SoC not using the common linker script?

Honestly I think that the error is not using the commen-arch linker script here, so my suggestion here is to proceed to fix this using the `ARCH_HAS_` in `arch/riscv/Kconfig`. @stephanosio any better idea here?

tari replied:
> Can you point me to the SoC not using the common linker script?

The major SoC that doesn't use the common one is `esp32c3` (it has a lot of unusual things in its linker script) but there were a few others I don't recall offhand.

carlocaione replied:
Ok, in general this is a SoC limitation. From the ARCH prospective nothing is preventing the usage of the code relocation. So IMO using `ARCH_HAS_CODE_DATA_RELOCATION` for all the RISCV boards is correct. `esp32c3` and others should simply not set `CODE_DATA_RELOCATION`.

tari replied:
Alright; changed the top-level `RISCV` config to select `ARCH_HAS_CODE_DATA_RELOCATOIN`.

tari commented:
> We need also to make this test-able by `tests/application_development/code_relocation/`

The test (and `gen_relocate_app`) had a few baked-in ARM assumptions that I've had to fix, but I've added a test case that passes for RISC-V. I had to make it a new test case since adding a new section requires some unportable configuration.

I'm not very happy with the way `target_relocation.cmake` currently assumes section names based on the target architecture, so ideas there are welcome. The best alternative I see is changing semantics slightly to drop the data-in-default-RAM optimization so we can remove any references to particular memory regions from the script.

carlocaione commented on cmake/linker/ld/target_relocation.cmake:
```diff
@@ -9,6 +9,14 @@ macro(toolchain_ld_relocation)
   set(MEM_RELOCATION_SRAM_BSS_LD
        "${PROJECT_BINARY_DIR}/include/generated/linker_sram_bss_relocate.ld")
   set(MEM_RELOCATION_CODE "${PROJECT_BINARY_DIR}/code_relocation.c")
+  if(CONFIG_ARM)
+    set(MEM_REGION_DEFAULT_RAM SRAM)
+  elseif(CONFIG_RISCV)
+    set(MEM_REGION_DEFAULT_RAM RAM)
+  else()
```
If you want you can submit a new PR for this but can we please unify the nomenclature between RISCV and ARM in the linker scripts instead of having `if-else` around?

tari replied:
Opened #49438 for that, which I will mark as ready for review once this is merged (since it builds on top of this branch).

carlocaione commented on tests/application_development/code_relocation/linker_riscv_qemu_sram2.ld:
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) ??? ???
```
?????? :)

tari replied:
That's what I get for deferring figuring out what I needed to fill in there! Updated to have an actual value.

carlocaione reviewed (changes requested):
The only blocking comment is the copyright. The other comment can be addressed here or in another PR.

edersondisouza commented on tests/application_development/code_relocation/linker_riscv_qemu_sram2.ld:
```diff
@@ -0,0 +1,34 @@
+ * @file
+ * @brief Linker command/script file
+ *
+ * Linker script for the Cortex-M platforms.
```
Cortex-M?

tari replied:
Hah, oops. I've added a commit to #49438 that fixes this.
